### PR TITLE
Move `DefaultIndexableGetter` to `Experimental::` and provide CTAD for `Indexables`

### PR DESCRIPTION
**Pull request by aprokop**
- Move `DefaultIndexableGetter` from `Details::` to `Experimental::`
- Provide CTAD for `Indexables`

### CODE REVIEW

**Comment by aprokop** [comment]
@dalg24 Thoughts on this PR?

**Review by dalg24 — COMMENTED** [review]
Please elaborate on the rational/motivation
As I recall it, in Boost that function object was about being able to extract the geometry from things like pair or tuples but here it looks like a lateral move and it is unclear to me what this improves over the status quo as currently proposed

**Comment by aprokop** [comment]
> Please elaborate on the rational/motivation

### Move to `Experimental::`

This is straightforward. We can't keep it in `Details` when it is part of the official interface and is a default argument to the constructor.

### Calling it `Indexable`

Better alignment with Boost.Geometry. Not a strong point, but there's little drawback for more compatibility.

### Templating it on value type

This strong types it, so that the indexable getter will be guaranteed to work with only the provided value type.
I'm less confident about that, and deferred expertise to Boost.

**Comment by dalg24** [comment]
> We can't keep it in `Details` when it is part of the official interface and is a default argument to the constructor.

Why not.  We can say there is a default and define clearly what it does without making it something public.  Users can deduce the type if they really want to spell it out.

**Comment by aprokop** [comment]
> Why not. We can say there is a default and define clearly what it does without making it something public. Users can deduce the type if they really want to spell it out.

If a user wants to specify a custom bounding volume type for the hierarchy, they would have to write
```c++
   BVH<MemorySpace, Values, ArborX::Details::DefaultIndexableGetter, BoundingVolume> bvh(blah...);
```
thus putting `Details::` namespace inside the user code.

**Comment by dalg24** [comment]
What benefit do we get from making it a template?  Were you planing on supporting `pair<Geometry, ...>` or whatnot?

**Comment by aprokop** [comment]
> What benefit do we get from making it a template? 

That is a good question. I don't have a good answer other than Boost does it. And I trust them to make good interfaces more than myself. Can you think of anything?

> Were you planing on supporting `pair<Geometry, ...>` or whatnot?

We already support `PairValueIndex`. In the previous version, it was one of the member functions, while in the new one it's partial specialization.

**Comment by aprokop** [comment]
I updated the PR. It no longer renames `DefaultIndexableGetter` and does not template it on the `Value`. Instead, it simply provides CTAD for `Indexables` and moves `DefaultIndexableGetter` from `Details` to `Experimental`.

File: src/spatial/detail/ArborX_IndexableGetter.hpp [hunk 1]
```diff
@@ -68,6 +76,16 @@ struct Indexables
...
 };
 
-} // namespace ArborX::Details
+template <typename Values, typename IndexableGetter>
+#if KOKKOS_VERSION >= 40400
+KOKKOS_DEDUCTION_GUIDE
+#else
+KOKKOS_FUNCTION
+#endif
+    Indexables(Values, IndexableGetter) -> Indexables<Values, IndexableGetter>;
```
Comment (dalg24) [inline src/spatial/detail/ArborX_IndexableGetter.hpp:85]: Please remind me why we need to provide a user-defined deduction guide and it does not work out of the box.

**Review by dalg24 — COMMENTED** [review]
Looks good, just wanting to confirm that the Indexables deduction guide is needed

File: src/spatial/detail/ArborX_IndexableGetter.hpp (hunk 1 above)
Comment (aprokop) [reply src/spatial/detail/ArborX_IndexableGetter.hpp:85]: Implicitly-generated deduction guide for aggregates is C++20.

**Review by dalg24 — APPROVED** [review]
I would guard the deduction guides with `#ifndef KOKKOS_ENABLE_CXX17` but leaving it up to you

**Pull request merged**